external-crates/move/crates/move-vm-runtime/src/execution/interpreter/locals.rs

```diff
@@ -0,0 +1,305 @@
+    pub fn take_loc(&mut self, ndx: BaseHeapId) -> PartialVMResult<Value> {
+        if self.is_invalid(ndx)? {
+            return Err(
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message("Cannot move from an invalid memory location".to_string()),
+            );
+        }
+
+        let Some(value_box) = self.values.get_mut(&ndx) else {
+            return Err(
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message(format!("Invalid index: {}", ndx)),
+            );
+        };
+
+        let value = std::mem::replace(value_box.as_mut(), Value::invalid());
+        Ok(value)
+    }
+
+    /// Borrows the specified location
+    pub fn borrow_loc(&self, ndx: BaseHeapId) -> PartialVMResult<Value> {
+        self.values
+            .get(&ndx)
+            .ok_or_else(|| {
+                PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR)
+                    .with_message(format!("Local index out of bounds: {}", ndx))
+            })
+            .and_then(|value| value.take_ref())
+    }
+
+    /// Checks if the value at the location is invalid
+    pub fn is_invalid(&self, ndx: BaseHeapId) -> PartialVMResult<bool> {
+        self.values
+            .get(&ndx)
+            .map(|value| matches!(value.as_ref(), &Value(ValueImpl::Invalid)))
+            .ok_or_else(|| {
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message(format!("Invalid index: {}", ndx))
+            })
+    }
+}
+
+// -------------------------------------------------------------------------------------------------
+// Machine (Runtime) Heap
+// -------------------------------------------------------------------------------------------------
+
+impl Default for MachineHeap {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl MachineHeap {
+    pub fn new() -> Self {
+        Self {
+        }
+    }
+
+    /// Allocates a stack frame with the given size.
+    /// If there is not enough space in the heap, it returns an error.
+    pub fn allocate_stack_frame(
+        &mut self,
+        params: Vec<Value>,
+        size: usize,
+    ) -> PartialVMResult<StackFrame> {
+        // Calculate how many invalid values need to be added
+        let invalids_len = size - params.len();
+
+        // Initialize the stack frame with the provided parameters and fill remaining slots with `Invalid`
+        let local_values = params
+            .into_iter()
+            .chain((0..invalids_len).map(|_| Value::invalid())) // Fill the rest with `Invalid`
+            .collect::<Vec<Value>>();
+
+        Ok(StackFrame { slice: local_values })
+    }
+
+    /// Frees the given stack frame, ensuring that it is the last frame on the heap.
+    pub fn free_stack_frame(&mut self, _frame: StackFrame) -> PartialVMResult<()> {
```

Comment:
Nit: may want to mark this (/maybe the definition of `MachineHeap`) with a TODO for greppability.

Reply:
TODO added